packages/flutter/lib/src/widgets/shortcuts.dart

```diff
@@ -93,9 +90,22 @@ class KeySet<T extends KeyboardKey> {
         && setEquals<T>(other._keys, _keys);
   }
 
+  // Cached hash code.
+  int _hashCode;
+
   @override
   int get hashCode {
-    return hashList(_keys);
+    // Keys are first sorted by hash to make sure the result of hashList is order independent.
+    if (_hashCode == null) {
```

Comment:
No, as far as I know, this isn't necessary.  `HashSet` is stable and not based on insertion order (`LinkedHashSet`, on the other hand is based on insertion order). Do you have an example that shows that it's not stable?

I switched to using a `HashSet` in https://github.com/flutter/flutter/pull/38936 to fix the bug I think you're trying to fix.

From the docs for `HashSet`:

"The iteration order of the set is not specified and depends on the hashcodes of the provided elements. However, the order is stable: multiple iterations over the same set produce the same order, as long as the set is not modified."

Reply:
The issue is that the insertion order is not what affects the computation of the hash code. What affects it is the order of elements we pass to `hashList`. While `LinkedHashSet` guarantees insertion order, `HashSet` does not guarantee any order. However, when you iterate over the elements of a `HashSet` it will return the elements in _some_ order. So when we pass it to `hashList` the elements are iterated in some unpredictable order.

Reply:
But a consistent order, based on the hash. Isn't that all we need?  I mean, your sorted version is just that: consistent and based on the hash.

Reply:
Hmm, perhaps the difference is that on the VM `HashSet` iteration order depends _only_ on the hashcodes of its elements while on the Web they do not. In general, AFAIK hash is not the only thing that determines the iteration order in a hash table. Two elements with different hashes can still end up in the same bucket, at which point there's no way to guarantee which of the two elements is extracted first. Let me poke at it for bit.

Reply:
Confirmed. Here's the difference between Web and VM:

```
Web:
set2: a (hash: 97) b (hash: 98) c (hash: 99) d (hash: 100)
set3: d (hash: 100) c (hash: 99) b (hash: 98) a (hash: 97)

VM:
set2: a (hash: 97) b (hash: 98) c (hash: 99) d (hash: 100)
set3: a (hash: 97) b (hash: 98) c (hash: 99) d (hash: 100)
```

This was printed using:

```dart
print(set2.keys.map((k) => '${k.keyLabel} (hash: ${k.hashCode})').join(' '));
print(set3.keys.map((k) => '${k.keyLabel} (hash: ${k.hashCode})').join(' '));
```

Reply:
So I think all the docs are saying is that the order depends on the hash, and that's it's stable across multiple iterations of the same set. But the docs do not say that hash is the only thing that determines the order, or that two sets containing the same elements will return them in the same order.

Reply:
Ahh, OK, but since on the VM things are iterated in hash order, maybe put an `if (kIsWeb)` around the sorting?  I'd rather not pay the price on non-web platforms if we can help it. There is a test that should fail if the VM implementation changes.

Reply:
I'm not sure the VM guarantees that, given what the docs say. We might have gotten lucky. But let me check, because the class is immutable the new implementation might be faster in both cases from caching the value.

Reply:
Ok, so I tried about 5 algorithms, and ran a benchmark for different key set sizes and different read/write ratios (i.e. how often `hashCode` is accessed per `KeySet` object). Caching the hash code value is strictly better. Even for low read/write ratio of 2 I got a 27% speedup in the worst case. The higher the ratio the higher the speedup. I got 927% improvement for key sets with 4 keys and read/write ratio of 32 ([benchmark results](https://docs.google.com/spreadsheets/d/1aSh6HbDp4mVoDN3Rn5jRT0Vtm0SDnQeNkNfsSu9kuaA/edit#gid=0)).

Reply:
That's excellent! The performance that is most important here is initial definition of the shortcuts map in the first place, as it affects initial build time (I had to re-land a more optimized shortcuts implementation because it exceeded our benchmarks there originally). How much slower is the sorted initial setup as opposed to just caching the hash value without sorting (only helpful for non-web cases, I realize).

Reply:
The initialization of the class has not changed. Are you worried about the existence of the extra `_hashCode` field? Trying to understand what benchmark I should run for this.

Reply:
It's the initialization of the maps that I'm concerned about, as those will require calling the `hashCode` function on each of the keys at least once. I know that if you use the pattern of putting the map inline in a `Shortcuts` widget `shortcuts` parameter, that it impacts build time.

For instance, here: https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/widgets/app.dart#L846

It's possible that since I moved it all to static arrays that compiler will have already calculated it, but I'm not sure.

I know it's probably not enough to worry about (we're talking some fairly small maps afterall), I was just stung by startup/build time when I originally added it.

Reply:
Ah, you mean the first access to `hashCode`. So, the numbers are as follows:

- 1-key sets *improved* by 16%
- 2-key sets *improved* by 0.84%
- 3-key sets *regressed* by 21%
- 4-key sets *regressed* by 20%

This is first access only. If the hash code is used a second time anywhere at all, it's offset immediately by the cache.

How many 3- and 4-key sets do we have in Flutter?

Reply:
Not many, and I wouldn't expect many.

Reply:
OK, let's go ahead with your changes.  Seems like a win, generally.

Reply:
Thanks! I'll keep an eye on the benchmarks for regressions.

Reply:
One thing that occurred to me: since you're now sorting things by hash, is it any faster to use the default Set implementation instead of HashSet? It might be able to be constructed faster.